src/Compilers/Core/Portable/ReferenceManager/CommonReferenceManager.Resolution.cs

```diff
@@ -746,13 +748,12 @@ private static void AddModule(PEModule module, int referenceIndex, ResolvedRefer
         protected void GetCompilationReferences(
             TCompilation compilation,
             DiagnosticBag diagnostics,
-            out ImmutableArray<MetadataReference> references,
-            out IDictionary<(string, string), MetadataReference> boundReferenceDirectives,
-            out ImmutableArray<Location> referenceDirectiveLocations)
+            out ImmutableArray<(MetadataReference MetadataReference, Location Location)> references,
```

Comment:
that was my initial feeling too but I ultimately I did it like this because [later on](https://github.com/dotnet/roslyn/pull/25422/files#diff-5be1efb173e01f50d763cf879b8408ffR234) (and [here](https://github.com/dotnet/roslyn/pull/25422/files#diff-5be1efb173e01f50d763cf879b8408ffR256) and a few other places down the road) the array length of all compilation `MetadataReferences` and all of these locations entries must be equal to each other - the existing resolution map code relies on matching indexes and a flattened structure is needed. As soon as the indexes would diverge, a much larger refactoring would be needed, which I wanted to avoid

if you want I can change the code to use `ImmutableArray<(ImmutableArray<MetadataReference> References, Location Location)>` here, but then my suggestion would still be to flatten it later on to avoid a big refactoring

Reply:
I pushed the change you suggested- it didn't affect things that much after all, we just iterate slightly differently now 👍